File: feishu_bot.py
"""
Lark 机器人集成（HTTP 推送模式）
- 使用 aiohttp 接收 Lark 事件推送
- 发送文本消息和卡片消息（用于确认按钮）
- 解析用户命令（/deploy, /cancel, /status）
"""

import asyncio
import logging
import json
import re
from typing import TYPE_CHECKING

import lark_oapi as lark
from lark_oapi.api.im.v1 import (
    CreateMessageRequest,
    CreateMessageRequestBody,
)
from aiohttp import web
import toml

if TYPE_CHECKING:
    from orchestrator import Orchestrator

logger = logging.getLogger(__name__)

_RE_DEPLOY = re.compile(r"^/deploy\s+(\S+)",        re.IGNORECASE)
_RE_CANCEL = re.compile(r"^/cancel\s+(\S+)\s*(.*)", re.IGNORECASE)
_RE_STATUS = re.compile(r"^/status$",               re.IGNORECASE)
_RE_AT     = re.compile(r"@\S+\s*")


class FeishuBot:
    """
    Lark 机器人。接收消息 → 调用 Orchestrator → 回复消息。

    支持的命令（在 Lark 群中 @机器人 发送）：
    - 普通文本          → 启动新任务
    - /deploy <task_id> → 确认部署
    - /cancel <task_id> → 取消任务
    - /status           → 查询当前任务状态
    """

    def __init__(self, config_path: str = "config.toml"):
        cfg = toml.load(config_path)
        feishu_cfg = cfg["feishu"]

        self._app_id     = feishu_cfg["app_id"]
        self._app_secret = feishu_cfg["app_secret"]
        self._port       = feishu_cfg.get("port", 8765)

        self._lark_client = (
            lark.Client.builder()
            .app_id(self._app_id)
            .app_secret(self._app_secret)
            .log_level(lark.LogLevel.WARNING)
            .build()
        )

        self._orchestrator: "Orchestrator | None" = None

    def set_orchestrator(self, orchestrator: "Orchestrator"):
        self._orchestrator = orchestrator

    # ------------------------------------------------------------------ #
    # 启动（HTTP 服务器）
    # ------------------------------------------------------------------ #

    def start(self):
        """启动 HTTP 事件服务器（阻塞）"""
        app = web.Application()
        app.router.add_post("/", self._handle_http_event)
        logger.info(f"[FeishuBot] 启动 HTTP 服务器，监听端口 {self._port}...")
        web.run_app(app, port=self._port, access_log=None)

    # ------------------------------------------------------------------ #
    # 事件处理
    # ------------------------------------------------------------------ #

    async def _handle_http_event(self, request: web.Request) -> web.Response:
        """处理 Lark HTTP 推送事件"""
        try:
            body = await request.json()
        except Exception:
            return web.Response(status=400)

        # URL 验证（首次配置时 Lark 发送 challenge）
        if body.get("type") == "url_verification":
            challenge = body.get("challenge", "")
            logger.info(f"[FeishuBot] URL 验证通过")
            return web.Response(
                text=json.dumps({"challenge": challenge}),
                content_type="application/json",
            )

        # 消息事件
        header = body.get("header", {})
        if header.get("event_type") == "im.message.receive_v1":
            asyncio.ensure_future(self._handle_message_event(body.get("event", {})))

        return web.Response(text="{}", content_type="application/json")

    async def _handle_message_event(self, event: dict):
        """解析消息事件并分发"""
        try:
            msg = event.get("message", {})

            if msg.get("message_type") != "text":
                return

            content_dict = json.loads(msg.get("content", "{}"))
            text = _RE_AT.sub("", content_dict.get("text", "")).strip()
            if not text:
                return

            chat_id    = msg.get("chat_id", "")
            message_id = msg.get("message_id", "")
            sender_id  = event.get("sender", {}).get("sender_id", {}).get("open_id", "")

            logger.info(f"[FeishuBot] 收到消息: chat={chat_id}, text={text[:80]}")

            await self._dispatch(chat_id, message_id, sender_id, text)
        except Exception as e:
            logger.exception(f"[FeishuBot] 消息处理异常: {e}")

    async def _dispatch(self, chat_id: str, message_id: str, sender_id: str, text: str):
        """根据文本内容分发到对应处理器"""
        if not self._orchestrator:
            logger.error("[FeishuBot] Orchestrator 未初始化")
            return

        m = _RE_DEPLOY.match(text)
        if m:
            await self._orchestrator.confirm_deploy(m.group(1), sender_id)
            return

        m = _RE_CANCEL.match(text)
        if m:
            await self._orchestrator.reject_task(m.group(1), m.group(2).strip())
            return

        if _RE_STATUS.match(text):
            active = self._orchestrator.list_active_tasks()
            if not active:
                await self.send_text(chat_id, "当前没有进行中的任务。")
            else:
                lines = ["**进行中的任务：**"]
                for t in active:
                    lines.append(f"• `{t.task_id}` [{t.state}]\n  {t.original_request[:60]}")
                await self.send_text(chat_id, "\n".join(lines))
            return

        await self._orchestrator.handle_feishu_message(chat_id, message_id, sender_id, text)

    # ------------------------------------------------------------------ #
    # 发送消息
    # ------------------------------------------------------------------ #

    async def send_text(self, chat_id: str, text: str):
        """发送纯文本消息"""
        try:
            request = (
                CreateMessageRequest.builder()
                .receive_id_type("chat_id")
                .request_body(
                    CreateMessageRequestBody.builder()
                    .receive_id(chat_id)
                    .msg_type("text")
                    .content(json.dumps({"text": text}))
                    .build()
                )
                .build()
            )
            loop = asyncio.get_event_loop()
            resp = await loop.run_in_executor(
                None, lambda: self._lark_client.im.v1.message.create(request)
            )
            if not resp.success():
                logger.error(f"[FeishuBot] 发送消息失败: {resp.msg}")
        except Exception as e:
            logger.exception(f"[FeishuBot] 发送消息异常: {e}")

    async def send_deploy_confirm_card(self, chat_id: str, task_id: str, summary: str):
        """发送部署确认交互卡片"""
        card = {
            "schema": "2.0",
            "body": {
                "elements": [
                    {
                        "tag": "div",
                        "text": {
                            "tag": "lark_md",
                            "content": f"**任务 `{task_id}` 需要确认部署**\n\n{summary}",
                        },
                    },
                    {
                        "tag": "action",
                        "actions": [
                            {
                                "tag": "button",
                                "text": {"tag": "plain_text", "content": "✅ 确认部署"},
                                "type": "primary",
                                "value": {"action": "deploy", "task_id": task_id},
                            },
                            {
                                "tag": "button",
                                "text": {"tag": "plain_text", "content": "❌ 取消"},
                                "type": "danger",
                                "value": {"action": "cancel", "task_id": task_id},
                            },
                        ],
                    },
                ]
            },
        }
        try:
            request = (
                CreateMessageRequest.builder()
                .receive_id_type("chat_id")
                .request_body(
                    CreateMessageRequestBody.builder()
                    .receive_id(chat_id)
                    .msg_type("interactive")
                    .content(json.dumps(card))
                    .build()
                )
                .build()
            )
            loop = asyncio.get_event_loop()
            resp = await loop.run_in_executor(
                None, lambda: self._lark_client.im.v1.message.create(request)
            )
            if not resp.success():
                logger.error(f"[FeishuBot] 发送卡片失败: {resp.msg}")
        except Exception as e:
            logger.exception(f"[FeishuBot] 发送卡片异常: {e}")

File: src/feishu_bot.ts
/**
 * 飞书机器人集成（HTTP 推送模式）
 * 用 @larksuiteoapi/node-sdk + express 替代 Python lark-oapi + aiohttp
 */

import express from 'express';
import * as lark from '@larksuiteoapi/node-sdk';
import { AppConfig } from './config.js';
import type { Orchestrator, NotifyPayload } from './orchestrator.js';

const RE_DEPLOY = /^\/deploy\s+(\S+)/i;
const RE_CANCEL = /^\/cancel\s+(\S+)\s*(.*)/i;
const RE_STATUS = /^\/status$/i;
const RE_AT     = /@\S+\s*/g;

export class FeishuBot {
  private cfg: AppConfig;
  private client: lark.Client;
  private orchestrator: Orchestrator | null = null;
  private port: number;

  constructor(cfg: AppConfig) {
    this.cfg = cfg;
    this.port = cfg.feishu.port;
    this.client = new lark.Client({
      appId: cfg.feishu.app_id,
      appSecret: cfg.feishu.app_secret,
      loggerLevel: lark.LoggerLevel.warn,
    });
  }

  setOrchestrator(orchestrator: Orchestrator): void {
    this.orchestrator = orchestrator;
  }

  // ── 启动 HTTP 服务器 ─────────────────────────────────────────────── //

  start(): void {
    const app = express();
    app.use(express.json());

    app.post('/', (req, res) => {
      void this.handleHttpEvent(req, res);
    });

    app.listen(this.port, () => {
      console.log(`[FeishuBot] HTTP 服务器启动，监听端口 ${this.port}`);
    });
  }

  // ── 事件处理 ─────────────────────────────────────────────────────── //

  private async handleHttpEvent(
    req: express.Request,
    res: express.Response,
  ): Promise<void> {
    try {
      const body = req.body as Record<string, unknown>;

      // URL 验证（握手阶段，跳过 token 校验）
      if (body['type'] === 'url_verification') {
        res.json({ challenge: body['challenge'] });
        return;
      }

      const header = body['header'] as Record<string, unknown> | undefined;

      // verification_token 校验（飞书在 header.token 中下发）
      if (this.cfg.feishu.verification_token) {
        const token = header?.['token'] as string | undefined;
        if (token !== this.cfg.feishu.verification_token) {
          console.warn('[FeishuBot] 非法请求：token 验证失败，已拒绝');
          res.status(403).json({});
          return;
        }
      }

      const eventType = header?.['event_type'] as string | undefined;

      // 消息事件
      if (eventType === 'im.message.receive_v1') {
        const event = body['event'] as Record<string, unknown>;
        this.handleMessageEvent(event).catch(e =>
          console.error(`[FeishuBot] 消息处理异常: ${e}`),
        );
        res.json({});
        return;
      }

      // 卡片按钮点击事件
      if (eventType === 'card.action.trigger') {
        const event = body['event'] as Record<string, unknown>;
        const actionValue = (event['action'] as Record<string, unknown> | undefined)
          ?.['value'] as { action?: string; task_id?: string } | undefined;
        const operator = event['operator'] as Record<string, unknown> | undefined;
        const userId = (operator?.['open_id'] as string | undefined) ?? '';

        if (actionValue?.action === 'deploy' && actionValue.task_id) {
          this.orchestrator?.confirmDeploy(actionValue.task_id, userId).catch(e =>
            console.error(`[FeishuBot] 卡片部署确认异常: ${e}`),
          );
          res.json({ toast: { type: 'success', content: '部署已确认，正在执行...' } });
        } else if (actionValue?.action === 'cancel' && actionValue.task_id) {
          this.orchestrator?.rejectTask(actionValue.task_id, '用户通过卡片取消').catch(e =>
            console.error(`[FeishuBot] 卡片取消异常: ${e}`),
          );
          res.json({ toast: { type: 'info', content: '任务已取消' } });
        } else {
          res.json({});
        }
        return;
      }

      res.json({});
    } catch (e) {
      console.error(`[FeishuBot] 事件处理异常: ${e}`);
      res.status(400).json({});
    }
  }

  private async handleMessageEvent(event: Record<string, unknown>): Promise<void> {
    try {
      const msg = event['message'] as Record<string, unknown>;
      if (msg['message_type'] !== 'text') return;

      const contentStr = msg['content'] as string ?? '{}';
      let contentDict: { text?: string } = {};
      try {
        contentDict = JSON.parse(contentStr) as { text?: string };
      } catch {
        console.warn(`[FeishuBot] 消息内容不是合法 JSON，已忽略: ${contentStr.slice(0, 100)}`);
        return;
      }
      const text = (contentDict.text ?? '').replace(RE_AT, '').trim();
      if (!text) return;

      const chatId    = msg['chat_id'] as string ?? '';
      const messageId = msg['message_id'] as string ?? '';
      const sender    = event['sender'] as Record<string, unknown> | undefined;
      const senderId  = (sender?.['sender_id'] as Record<string, string> | undefined)?.['open_id'] ?? '';

      console.log(`[FeishuBot] 收到消息: chat=${chatId}, text=${text.slice(0, 80)}`);
      await this.dispatch(chatId, messageId, senderId, text);
    } catch (e) {
      console.error(`[FeishuBot] 消息处理异常: ${e}`);
    }
  }

  private async dispatch(
    chatId: string,
    messageId: string,
    senderId: string,
    text: string,
  ): Promise<void> {
    if (!this.orchestrator) {
      console.error('[FeishuBot] Orchestrator 未初始化');
      return;
    }

    let m: RegExpMatchArray | null;

    m = text.match(RE_DEPLOY);
    if (m) {
      await this.orchestrator.confirmDeploy(m[1], senderId);
      return;
    }

    m = text.match(RE_CANCEL);
    if (m) {
      await this.orchestrator.rejectTask(m[1], m[2].trim());
      return;
    }

    if (RE_STATUS.test(text)) {
      const active = this.orchestrator.listActiveTasks();
      if (active.length === 0) {
        await this.sendText(chatId, '当前没有进行中的任务。');
      } else {
        const lines = ['**进行中的任务：**'];
        for (const t of active) {
          lines.push(`• \`${t.task_id}\` [${t.state}]\n  ${t.original_request.slice(0, 60)}`);
        }
        await this.sendText(chatId, lines.join('\n'));
      }
      return;
    }

    await this.orchestrator.handleFeishuMessage(chatId, messageId, senderId, text);
  }

  // ── NotifyCallback：Orchestrator 调用的统一通知入口 ─────────────── //

  async handleNotify(chatId: string, payload: NotifyPayload): Promise<void> {
    if (payload.kind === 'text') {
      await this.sendText(chatId, payload.text);
    } else {
      await this.sendDeployConfirmCard(chatId, payload.taskId, payload.summary);
    }
  }

  // ── 发送消息 ─────────────────────────────────────────────────────── //

  private async sendMessage(chatId: string, msgType: string, content: unknown): Promise<void> {
    await this.client.im.v1.message.create({
      params: { receive_id_type: 'chat_id' },
      data: {
        receive_id: chatId,
        msg_type: msgType,
        content: JSON.stringify(content),
      },
    });
  }

  async sendText(chatId: string, text: string): Promise<void> {
    try {
      await this.sendMessage(chatId, 'text', { text });
    } catch (e) {
      console.error(`[FeishuBot] 发送消息异常: ${e}`);
    }
  }

  async sendDeployConfirmCard(chatId: string, taskId: string, summary: string): Promise<void> {
    const card = {
      schema: '2.0',
      body: {
        elements: [
          {
            tag: 'div',
            text: {
              tag: 'lark_md',
              content: `**任务 \`${taskId}\` 需要确认部署**\n\n${summary}`,
            },
          },
          {
            tag: 'action',
            actions: [
              {
                tag: 'button',
                text: { tag: 'plain_text', content: '✅ 确认部署' },
                type: 'primary',
                value: { action: 'deploy', task_id: taskId },
              },
              {
                tag: 'button',
                text: { tag: 'plain_text', content: '❌ 取消' },
                type: 'danger',
                value: { action: 'cancel', task_id: taskId },
              },
            ],
          },
        ],
      },
    };

    try {
      await this.sendMessage(chatId, 'interactive', card);
    } catch (e) {
      console.error(`[FeishuBot] 发送卡片异常: ${e}`);
      // 降级为文本
      await this.sendText(
        chatId,
        `✅ [${taskId}] 测试通过！\n${summary}\n\n• 确认部署：\`/deploy ${taskId}\`\n• 取消：\`/cancel ${taskId}\``,
      );
    }
  }
}
